pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -317,10 +307,10 @@ impl InMemoryLayer {
     /// Records the end_lsn for non-dropped layers.
     /// `end_lsn` is exclusive
     pub fn freeze(&self, end_lsn: Lsn) {
-        let mut inner = self.inner.write().unwrap();
+        let inner = self.inner.write().unwrap();
 
         assert!(self.start_lsn < end_lsn);
-        inner.end_lsn = Some(end_lsn);
+        assert_eq!(self.end_lsn.fetch_max(end_lsn), Lsn::INVALID);
```

Comment:
I think that is equivalent to this mutex-pseudo:

```rust
let prev = {
  let g = self.end_lsn.lock().unwrap();
  let prev = *g;
  *g = std::cmp::max(prev, end_lsn);
};

// this is the once assumption, which is why I suggested OnceCell
assert_eq!(prev, Lsn::INVALID);
```

The atomic wrapper fns come really ackward without seeing the Ordering, though, I guess it would not make too clear that it will do be a store as well unless `Ordering::Release` was used instead of what is now used `Ordering::AcqRel`.

Reply:
Yeah `fetch_max` could be named in a better way I guess, but it's [named the same way](https://doc.rust-lang.org/nightly/std/sync/atomic/struct.AtomicU64.html#method.fetch_max) on the stdlib atomics, and cpp uses the same naming scheme. It is also named that way [outside of cpp](https://en.wikipedia.org/wiki/Fetch-and-add).

I'm using a `OnceLock` now instead, so this should be resolved.

Reply:
I'd consider the `fetch_max` to be read only if I'd see `Ordering::Acquire` next to it, but yeah, the naming ship has sailed :)